mesh/ncmesh.hpp

```diff
@@ -673,17 +701,22 @@ class NCMesh
    Array<Refinement> ref_stack; ///< stack of scheduled refinements (temporary)
    HashTable<Node> shadow; ///< temporary storage for reparented nodes
    Array<Triple<int, int, int> > reparents; ///< scheduled node reparents (tmp)
+   Array<real_t> reparent_scale;  ///< scale associated with reparents (tmp)
 
    Table derefinements; ///< possible derefinements, see GetDerefinementTable
 
    /** Refine the element @a elem with the refinement @a ref_type
-       (c.f. Refinement::enum) */
-   void RefineElement(int elem, char ref_type);
+       (c.f. Refinement::enum) and scale in each direction. */
+   void RefineElement(int elem, char ref_type, real_t scale_x = 0.5,
+                      real_t scale_y = 0.5, real_t scale_z = 0.5);
```

Comment:
The reason for the awkward input of 3 scalings is to preserve backward compatibility. I think the most sensible interface would be
`void RefineElement(const Refinement & ref)`
since `struct Refinement` now contains the element index, refinement type, and 3 scalings. This could be added as a new function, but the old function needs to be maintained. We can have the old and new functions, but some function has to contain the actual code, while the other just calls that function. The code in `RefineElement` actually uses the 5 input variables explicitly, so it is convenient to keep them as separate variables. For example, if we only input `Refinement & ref`, we would still want things like `scale_z = ref.s[2]`. So we could add a new function with the single argument `Refinement & ref`, but it would simply wrap the existing function, and I don't see that simplifying much. The API would be cleaner if we just replace the 5 arguments with `Refinement & ref`, but that breaking change would need to be done in MFEM 5.0. I think you have a good suggestion, but I'm not sure it can really help until 5.0.

Reply:
> So we could add a new function with the single argument Refinement & ref, but it would simply wrap the existing function, and I don't see that simplifying much.

I see your point about effectively wrapping this function with `RefineElement(const Refinement &ref)` and then having `RefineElement(int elem, char ref_type, real_t scale=0.5)` call the former to keep backwards compatibility, but I personally prefer that API over having to, for example, specify dummy values for `scale_x` and `scale_y` for Z refinement... I find it nicely provides separate functions for isotropic and anisotropic refinement, even if one simply calls the other.

All that said, this is my subjective opinion and there's already one approval for this PR, so I don't want to hold it this important functionality over this (i.e., I'll approve once the indentation issues are fixed).

Reply:
I added this wrapping in c03257233f58c2c78cfea963eaa520c627bf2dc1. For this PR, it only simplifies one line (ncmesh.cpp:1943), but I agree it is good to add the new function for future use.

Reply:
Unfortunately, even this new function does not eliminate the 3 scaling parameters that you would like to avoid. I don't see a good way to do that without breaking backward compatibility. My hope is that in the future, we can just remove that function and keep the new one, which is a breaking change.

Reply:
I restored the old interface, even without a single `scale` argument, in 4b87236e738c15b867faeb2277911f1efa50424e. The impact was much less than I expected, so the code remains very simple with this change.

Reply:
This looks great!  Thanks for doing all the iterations!